keps/sig-architecture/2019-07-16-node-role-label-use.md

```diff
@@ -0,0 +1,211 @@
+
+Therefore, for each change we recommend the following process to adopt the new labels in successive releases:
+
+* Release 1 (1.16):
+  * Introduce a feature gate for disabling node-role being honored. The gate defaults to on. `LegacyNodeRoleBehavior=on`
+  * Define the new node label with an associated feature gate for each feature area. The gate defaults to off. `FeatureA=off`
+  * Behavior for each functional area is defined as `LegacyNodeRoleBehavior == on && node_has_role || FeatureA == on && node_has_label`
+  * Deprecation of node role behavior in tree is announced for a future release as a warning to users
+  * No new components may leverage node-roles within Kubernetes projects.
+  * Early adopters may label their nodes to opt in to the features, even in the absence of the  gate.
+* Release 2 (1.17):
+  * For each new node label, usage is reviewed and as appropriate the label is declared beta/GA and the feature gate is set on
+  * All Kubernetes deployments should be updated to add node labels as appropriate: `kubectl label nodes -l node-role.kubernetes.io/master LABEL_A=VALUE_A`
+  * Documentation will be provided on making the transition
+  * Deployments may set `LegacyNodeRoleBehavior=off` after they have set the appropriate labels.
+  * NOTE: Release 3 starts when all labels graduate to beta
+* Release 3 (1.18):
+  * Default the `LegacyNodeRoleBehavior=off`. Admins whose deployments still use the old labels may set `LegacyNodeRoleBehavior=on` during 1.17 to get the legacy behavior.
+  * Deployments should stop setting `LegacyNodeRoleBehavior=off` if they opted out early.
+* Release 4 (1.19):
+  * The `LegacyNodeRoleBehavior` gate and all feature-level gates are removed, components that attempt to set these gates will fail to start.
+  * Code that references node-roles within Kubernetes will be removed.
+
+In Release 4 (which could be as early as 1.19) this KEP will be considered complete.
+
+
+### Current users of `node-role.kubernetes.io/*` within the project that must change
+
+The following components vary behavior based on the presence of the node-role labels:
+
+
+#### Service load-balancer
+
+The service load balancer implementation previously implemented a heuristic where `node-role.kubernetes.io/master` is used to exclude masters from the candidate nodes for a service. This is an implementation detail of the cluster and is not allowed. Since there is value in excluding nodes from service load balancer candidacy in some deployments, an alpha feature gated label `alpha.service-controller.kubernetes.io/exclude-balancer` was added in Kubernetes 1.9.
+
+This label should be moved to beta in Kube 1.16 at its final name `node.kubernetes.io/exclude-balancer`, its feature gate `ServiceNodeExclusion` should default on in 1.17, the gate `ServiceNodeExclusion` should be declared GA in 1.18, and the gate will be removed in 1.19. The old alpha label should be honored in 1.16 and removed in 1.17.
```

Comment:
I know this is an alpha feature, but any reason we wouldn't honor the alpha label longer than 1.16? Code wise it shouldn't be much to maintain and the risks for a cluster not replacing the alpha for v1.17 seems high.

Reply:
agree, no pressure to remove it other than tidiness